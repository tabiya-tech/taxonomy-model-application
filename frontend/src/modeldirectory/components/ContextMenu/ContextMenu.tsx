import * as React from "react";
import { ListItemIcon, ListItemText, Menu, MenuItem, Typography } from "@mui/material";
import CloudDownloadIcon from "@mui/icons-material/CloudDownload";

export interface ContextMenuProps {
  anchorEl?: HTMLElement | null;
  open: boolean;
  isExportDisabled: boolean;
  notifyOnClose: () => void;
  notifyOnExport: () => void;
}

const uniqueId = "715a535d-ca4b-4252-b967-8d5b5b2381b9";

export const DATA_TEST_ID = {
  CONTEXT_MENU: `context-menu-${uniqueId}`,
  MENU_ITEM_EXPORT: `menu-item-export-${uniqueId}`,
};

const ContextMenu = (props: Readonly<ContextMenuProps>) => {
  function handleExport() {
    props.notifyOnClose();
    props.notifyOnExport();
  }

  return (
    <Menu
      data-testid={DATA_TEST_ID.CONTEXT_MENU}
      elevation={2}
      anchorOrigin={{
        vertical: "bottom",
        horizontal: "center",
      }}
      transformOrigin={{
        vertical: "top",
        horizontal: "right",
      }}
      anchorEl={props.anchorEl}
      open={props.open}
      onClose={props.notifyOnClose}
    >
      <MenuItem
        onClick={() => handleExport()}
        component="button"
        data-testid={DATA_TEST_ID.MENU_ITEM_EXPORT}
        disabled={props.isExportDisabled}
      >
        <ListItemIcon>
          <CloudDownloadIcon color="primary" />
        </ListItemIcon>
        <ListItemText>
          <Typography variant={"caption"} color={"primary"}>
            Export
          </Typography>
        </ListItemText>
      </MenuItem>
    </Menu>
  );
};

export default ContextMenu;
