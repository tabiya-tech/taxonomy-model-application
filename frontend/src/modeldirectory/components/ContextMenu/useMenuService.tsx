import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import ModelInfo = ModelInfoTypes.ModelInfo;
import React, { useContext, useEffect, useState } from "react";
import { IsOnlineContext } from "src/app/providers";

export const isModelNotSuccessfullyImported = (model: ModelInfo) => {
  return (
    model?.importProcessState?.status !== ImportProcessStateAPISpecs.Enums.Status.COMPLETED ||
    model?.importProcessState?.result.errored !== false
  );
};

export function useMenuService() {
  const isOnline = useContext(IsOnlineContext);
  const [menuState, setMenuState] = useState<{
    anchorEl: HTMLElement | null;
    open: boolean;
    isExportDisabled: boolean;
  }>({
    anchorEl: null,
    open: false,
    isExportDisabled: !isOnline,
  });
  const [model, setModel] = useState<ModelInfo | null>(null);

  const openMenu = (event: React.MouseEvent<HTMLButtonElement, MouseEvent>, model: ModelInfo) => {
    setMenuState({
      anchorEl: event.currentTarget,
      open: true,
      isExportDisabled: !isOnline || isModelNotSuccessfullyImported(model),
    });
    setModel(model);
  };

  const closeMenu = () => {
    setMenuState({
      anchorEl: null,
      open: false,
      isExportDisabled: !isOnline,
    });
    setModel(null);
  };

  useEffect(() => {
    if (!isOnline) {
      setMenuState((prevState) => ({
        ...prevState,
        isExportDisabled: true,
      }));
    } else {
      setMenuState((prevState) => ({
        ...prevState,
        isExportDisabled: isModelNotSuccessfullyImported(model!),
      }));
    }
  }, [isOnline, model]);

  return {
    menuState,
    model,
    openMenu,
    closeMenu,
  };
}
