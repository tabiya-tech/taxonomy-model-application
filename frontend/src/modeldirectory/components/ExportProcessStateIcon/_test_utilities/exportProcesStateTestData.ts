import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import ExportProcessStateAPISpecs from "api-specifications/importProcessState";
import { getMockId } from "src/_test_utilities/mockMongoId";

export const getAllExportProcessStatePermutations = (): ModelInfoTypes.ExportProcessState[] => {
  // generate all permutations of ExportProcessState status, errored, exportErrors, exportWarnings
  const allStatuses = Object.values(ExportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  const allBooleans = [true, false];

  const allPermutations: ModelInfoTypes.ExportProcessState[] = [];
  allStatuses.forEach((status) => {
    allBooleans.forEach((errored) => {
      allBooleans.forEach((exportErrors) => {
        allBooleans.forEach((exportWarnings) => {
          allPermutations.push({
            id: getMockId(allPermutations.length + 1),
            status,
            result: {
              errored,
              exportErrors: exportErrors,
              exportWarnings: exportWarnings,
            },
            downloadUrl: "https://example.com",
            timestamp: new Date(),
          });
        });
      });
    });
  });
  return allPermutations;
};
