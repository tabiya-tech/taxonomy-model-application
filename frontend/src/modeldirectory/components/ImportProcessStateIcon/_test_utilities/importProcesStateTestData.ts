import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import { getMockId } from "src/_test_utilities/mockMongoId";

export const getAllImportProcessStatePermutations = (): ModelInfoTypes.ImportProcessState[] => {
  // generate all permutations of ImportProcessState status, errored, parsingErrors, parsingWarnings
  const allStatuses = Object.values(ImportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  const allBooleans = [true, false];

  const allPermutations: ModelInfoTypes.ImportProcessState[] = [];
  allStatuses.forEach((status) => {
    allBooleans.forEach((errored) => {
      allBooleans.forEach((parsingErrors) => {
        allBooleans.forEach((parsingWarnings) => {
          allPermutations.push({
            id: getMockId(allPermutations.length + 1),
            status,
            result: {
              errored,
              parsingErrors: parsingErrors,
              parsingWarnings: parsingWarnings,
            },
            createdAt: new Date(allPermutations.length),
            updatedAt: new Date(allPermutations.length),
          });
        });
      });
    });
  });
  return allPermutations;
};
