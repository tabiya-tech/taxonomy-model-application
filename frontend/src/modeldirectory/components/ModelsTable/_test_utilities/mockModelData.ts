import ModelInfoAPISpecs from "api-specifications/modelInfo";
import LocaleAPISpecs from "api-specifications/locale";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import ExportProcessStateAPISpecs from "api-specifications/exportProcessState";
import { ImportProcessStateEnums } from "api-specifications/importProcessState/enums";

import { v4 as uuidv4 } from "uuid";

import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import { getMockId } from "src/_test_utilities/mockMongoId";
import { getRandomLorem, getRandomString, getTestString } from "src/_test_utilities/specialCharacters";
import { faker } from "@faker-js/faker";
import { CELL_MAX_LENGTH } from "src/modeldirectory/components/ModelsTable/ModelsTable";

export function getOneFakeModel(id?: number): ModelInfoTypes.ModelInfo {
  const model = getArrayOfFakeModels(1)[0];
  model.id = getMockId(id ?? 1);
  return model;
}

export function getArrayOfFakeModels(count: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < count; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: `${i + 1}/${count} - ${getRandomLorem(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH)}`.substring(
        0,
        ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH
      ),
      locale: {
        UUID: uuidv4(),
        name: faker.location.country().substring(0, ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: faker.location
          .countryCode("alpha-3")
          .substring(0, LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description:
        i % 2 === 0
          ? getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH)
          : getRandomLorem(CELL_MAX_LENGTH / 2), // 50% chance of long description
      released: i % 2 === 0, // 50% chance of released
      releaseNotes: faker.lorem.text().substring(0, ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: faker.system.semver().substring(0, ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export function getArrayOfFakeModelsMaxLength(count: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < count; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: `${i + 1}/${count} - ${getRandomLorem(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH)}`.substring(
        0,
        ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH
      ),
      locale: {
        UUID: uuidv4(),
        name: getRandomLorem(LocaleAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: getRandomLorem(LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description: getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
      released: i % 2 === 0, // 50% chance of released
      releaseNotes: getRandomLorem(ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getRandomLorem(ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export function getOneRandomModelMaxLength(): ModelInfoTypes.ModelInfo {
  return getArrayOfRandomModelsMaxLength(1)[0];
}

export function getArrayOfRandomModelsMaxLength(number: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < number; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: getRandomString(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
      locale: {
        UUID: uuidv4(),
        name: getRandomString(LocaleAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: getTestString(LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description: getTestString(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
      released: i % 2 === 0,
      releaseNotes: getTestString(ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getTestString(ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export const fakeModel: ModelInfoTypes.ModelInfo = {
  UUID: "8d914eab-6f7d-4183-accc-a09f99887b39",
  UUIDHistory: ["8d914eab-6f7d-4183-accc-a09f99887b39"],
  description: "aw j zt   agrkasl dy ogtimpsauwumu l utrovthao syertm beawpxluhyudgzbbm",
  id: "000000000000000000000001",
  importProcessState: {
    id: "000000000000000000000001",
    result: {
      errored: true,
      parsingErrors: true,
      parsingWarnings: true,
    },
    status: ImportProcessStateEnums.Status.PENDING,
    createdAt: new Date("2023-10-18T17:35:10.571Z"),
    updatedAt: new Date("2023-10-18T17:35:12.571Z"),
  },
  exportProcessState: [
    {
      downloadUrl: "https://www.example.com/",
      id: "000000000000000000000001",
      result: {
        errored: true,
        exportErrors: true,
        exportWarnings: true,
      },
      status: ExportProcessStateAPISpecs.Enums.Status.PENDING,
      timestamp: new Date("2023-10-18T17:35:10.571Z"),
      createdAt: new Date("2023-10-18T17:35:10.571Z"),
      updatedAt: new Date("2023-10-18T17:35:12.571Z"),
    },
  ],
  locale: {
    UUID: "7c4ea00b-fbea-4a31-9c04-009c813f5f34",
    name: "Palestine",
    shortCode: "ETH",
  },
  name: "1/1 - fysmros tcgcnjbbrrev",
  path: "https://unwelcome-editorial.net/",
  releaseNotes: "Perferendis modi impedit necessitatibus a",
  released: true,
  tabiyaPath: "https://silver-vulture.biz",
  createdAt: new Date("2023-10-18T17:35:10.571Z"),
  updatedAt: new Date("2022-10-18T17:35:10.571Z"),
  version: "0.2.2",
};

export function getRandomImportStatus(id: number) {
  const allStatuses = Object.values(ImportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  return allStatuses[id % allStatuses.length];
}

export function getRandomExportStatus(id: number) {
  const allStatuses = Object.values(ExportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  return allStatuses[id % allStatuses.length];
}

export function getOneFakeExportProcessState(i: number): ModelInfoTypes.ExportProcessState {
  return {
    id: getMockId(10000 + i),
    status: getRandomExportStatus(i),
    result: {
      errored: i % 2 === 0,
      exportErrors: i % 2 === 0,
      exportWarnings: i % 2 === 0,
    },
    downloadUrl: faker.internet.url(),
    timestamp: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    updatedAt: new Date(),
  };
}

export function getOneFakeImportProcessState(i: number): ModelInfoTypes.ImportProcessState {
  return {
    id: getMockId(10000 + i),
    status: getRandomImportStatus(i),
    result: {
      errored: i % 2 === 0,
      parsingErrors: i % 2 === 0,
      parsingWarnings: i % 2 === 0,
    },
    createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    updatedAt: new Date(),
  };
}
