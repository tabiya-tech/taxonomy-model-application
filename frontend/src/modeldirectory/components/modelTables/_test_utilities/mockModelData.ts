import ModelInfoAPISpecs from "api-specifications/modelInfo";
import LocaleAPISpecs from "api-specifications/locale";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import {v4 as uuidv4} from "uuid";

import {ModelInfoTypes} from "src/modelInfo/modelInfoTypes";
import {getMockId} from "src/_test_utilities/mockMongoId";
import {getRandomLorem, getRandomString, getTestString} from "src/_test_utilities/specialCharacters";
import {faker} from "@faker-js/faker";
import {CELL_MAX_LENGTH} from "../ModelsTable";

export function getOneFakeModel(id?: number): ModelInfoTypes.ModelInfo {
  return getArrayOfFakeModels(id ? id : 1)[0];
}

export function getArrayOfFakeModels(number: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < number; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      previousUUID: uuidv4(),
      originUUID: uuidv4(),
      name: Array.from({length: 20}, (_, i) => faker.lorem.word()).join(" ").substring(0, ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
      locale: {
        UUID: uuidv4(),
        name: faker.location.country().substring(0, ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: faker.location.countryCode("alpha-3").substring(0, LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH)
      },
      description: i % 2 === 0 ? getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH) : getRandomLorem(CELL_MAX_LENGTH / 2), // 50% chance of long description
      released: i % 2 === 0, // 50% chance of released
      releaseNotes: faker.lorem.text().substring(0, ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: faker.system.semver().substring(0, ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      importProcessState: {
        id: getMockId(10000 + i),
        status: getRandomStatus(i),
        result: {
          errored: i % 2 === 0,
          parsingErrors: i % 2 === 0,
          parsingWarnings: i % 2 === 0,
        }
      }
    };
    models.push(model);
  }
  return models
}

export function getArrayOfFakeModelsMaxLength(number: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < number; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      previousUUID: uuidv4(),
      originUUID: uuidv4(),
      name: getRandomLorem(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
      locale: {
        UUID: uuidv4(),
        name: getRandomLorem(LocaleAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: getRandomLorem(LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH)
      },
      description: getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
      released: i % 2 === 0, // 50% chance of released
      releaseNotes: getRandomLorem(ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getRandomLorem(ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      importProcessState: {
        id: getMockId(10000 + i),
        status: getRandomStatus(i),
        result: {
          errored: i % 2 === 0,
          parsingErrors: i % 2 === 0,
          parsingWarnings: i % 2 === 0,
        }
      }
    };
    models.push(model);
  }
  return models
}

export function getOneRandomModelMaxLength(): ModelInfoTypes.ModelInfo {
  return getArrayOfRandomModelsMaxLength(1)[0];
}

export function getArrayOfRandomModelsMaxLength(number: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < number; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      previousUUID: uuidv4(),
      originUUID: uuidv4(),
      name: getRandomString(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
      locale: {
        UUID: uuidv4(),
        name: getRandomString(LocaleAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: getTestString(LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH)
      },
      description: getTestString(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
      released: i % 2 === 0,
      releaseNotes: getTestString(ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getTestString(ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      importProcessState: {
        id: getMockId(10000 + i),
        status: getRandomStatus(i),
        result: {
          errored: i % 2 === 0,
          parsingErrors: i % 2 === 0,
          parsingWarnings: i % 2 === 0,
        }
      }
    };
    models.push(model);
  }
  return models
}

export function getArrayOfFakeModelsForSorting(): ModelInfoTypes.ModelInfo[] {
  const baseModel: Omit<ModelInfoTypes.ModelInfo, 'name' | 'version' | 'createdAt' | 'updatedAt'> = {
    id: 'base',
    UUID: uuidv4(),
    previousUUID: uuidv4(),
    originUUID: uuidv4(),
    locale: {
      UUID: uuidv4(),
      name: faker.location.country().substring(0, ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
      shortCode: faker.location.countryCode("alpha-3").substring(0, LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH)
    },
    description: getRandomLorem(CELL_MAX_LENGTH / 2),
    released: true,
    releaseNotes: faker.lorem.text().substring(0, ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
    path: faker.internet.url(),
    tabiyaPath: faker.internet.url(),
    importProcessState: {
      id: getMockId(10000),
      status: getRandomStatus(1),
      result: {
        errored: true,
        parsingErrors: true,
        parsingWarnings: true,
      }
    }
  };

  return [
    {
      ...baseModel,
      id: '1',
      name: 'ancient',
      version: '1.0.0',
      createdAt: new Date('2022-01-01T00:00:00.000Z'),
      updatedAt: new Date('2022-01-01T00:00:00.000Z'),
    },
    {
      ...baseModel,
      id: '2',
      name: 'older',
      version: '1.1.0',
      createdAt: new Date('2022-01-02T00:00:00.000Z'),
      updatedAt: new Date('2022-01-02T00:00:00.000Z'),
    },
    {
      ...baseModel,
      id: '3',
      name: 'old',
      version: '1.1.1',
      createdAt: new Date('2022-01-03T00:00:00.000Z'),
      updatedAt: new Date('2022-01-03T00:00:00.000Z'),
    },
    {
      ...baseModel,
      id: '4',
      name: 'new',
      version: '1.1.2',
      createdAt: new Date('2022-01-04T00:00:00.000Z'),
      updatedAt: new Date('2022-01-04T00:00:00.000Z'),
    },
    {
      ...baseModel,
      id: '5',
      name: 'newer',
      version: '1.1.3',
      createdAt: new Date('2022-01-05T00:00:00.000Z'),
      updatedAt: new Date('2022-01-05T00:00:00.000Z'),
    },
    {
      ...baseModel,
      id: '6',
      name: 'newest',
      version: '1.2.2',
      createdAt: new Date('2022-01-06T00:00:00.000Z'),
      updatedAt: new Date('2022-01-06T00:00:00.000Z'),
    }
  ];
}


export function getRandomStatus(id: number) {
  const allStatuses = Object.values(ImportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  return allStatuses[id % allStatuses.length];
}
