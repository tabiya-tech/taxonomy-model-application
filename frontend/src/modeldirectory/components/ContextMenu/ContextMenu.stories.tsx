import { action } from "@storybook/addon-actions";
import { Meta, StoryObj } from "@storybook/react";
import ContextMenu, { ContextMenuProps } from "./ContextMenu";
import * as React from "react";
import { useEffect } from "react";
import Box from "@mui/material/Box";
import { getOneRandomModelMaxLength } from "src/modeldirectory/components/modelTables/_test_utilities/mockModelData";

const meta: Meta<typeof ContextMenu> = {
  title: "ModelDirectory/ModelsTable/ContextMenu",
  component: ContextMenu,
  tags: ["autodocs"],
  argTypes: {
    open: {
      control: {
        type: "boolean",
      },
    },
    notifyOnClose: {
      table: {
        disable: true,
      },
    },
  },
};

export default meta;
type Story = StoryObj<typeof ContextMenu>;

export const Shown: Story = {
  render: (args) => <SetupComponent {...args} />,
  args: {
    open: true,
  },
};

function SetupComponent(props: Readonly<ContextMenuProps>) {
  const [anchorEl, setAnchorEl] = React.useState<null | HTMLElement>(null);
  useEffect(() => {
    const anchor = document.getElementById("anchor-on-me");
    setAnchorEl(anchor);
  }, []);

  return (
    <>
      <Box display="flex" justifyContent="center" alignItems="center" height="100%">
        <div style={{ border: "1px dashed", height: "20px", width: "100px" }} id="anchor-on-me" />
      </Box>
      <ContextMenu
        anchorEl={anchorEl}
        model={getOneRandomModelMaxLength()}
        open={anchorEl !== null && props.open}
        notifyOnClose={action("notifyOnClose")}
      />
    </>
  );
}
