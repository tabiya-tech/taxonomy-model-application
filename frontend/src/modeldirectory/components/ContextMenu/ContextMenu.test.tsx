// mute the console
import "src/_test_utilities/consoleMock";

import { fireEvent, render, waitFor, screen } from "@testing-library/react";
import ContextMenu, { ContextMenuProps, DATA_TEST_ID } from "./ContextMenu";
import { getOneRandomModelMaxLength } from "src/modeldirectory/components/ModelsTable/_test_utilities/mockModelData";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";

describe("ContextMenu", () => {
  const stdGivenProps: ContextMenuProps = {
    anchorEl: document.createElement("div"), // Mock an HTMLElement
    model: getOneRandomModelMaxLength(),
    open: true,
    notifyOnClose: jest.fn(),
    notifyOnExport: jest.fn(),
  };

  describe("Render tests", () => {
    beforeEach(() => {
      jest.clearAllMocks();
    });
    test("should render the ContextMenu visible", () => {
      // WHEN the ContextMenu is rendered with the std given props
      render(<ContextMenu {...stdGivenProps} />);

      // THEN expect no errors or warning to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();
      // AND the ContextMenu to be visible
      const actualMenu = screen.getByTestId(DATA_TEST_ID.CONTEXT_MENU);
      expect(actualMenu).toBeInTheDocument();
      // AND to match the snapshot
      expect(actualMenu).toMatchSnapshot();
      // AND the export menu item to be visible
      expect(screen.getByTestId(DATA_TEST_ID.MENU_ITEM_EXPORT)).toBeInTheDocument();
    });

    test("should render the ContextMenu hidden", () => {
      // WHEN the ContextMenu is rendered closed with the std given props
      render(<ContextMenu {...stdGivenProps} open={false} />);

      // THEN expect no errors or warning to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();
      // AND the ContextMenu to be hidden
      const actualMenu = screen.queryByTestId(DATA_TEST_ID.CONTEXT_MENU);
      expect(actualMenu).not.toBeInTheDocument();
      // AND the export menu item to not be shown
      expect(screen.queryByTestId(DATA_TEST_ID.MENU_ITEM_EXPORT)).not.toBeInTheDocument();
    });
  });

  describe("Action tests", () => {
    beforeEach(() => {
      jest.resetAllMocks();
    });

    test("should call notifyOnExport and notifyOnClose when the export menu item is clicked", async () => {
      // WHEN the ContextMenu is rendered with the std given props
      render(
        <ContextMenu
          {...stdGivenProps}
          model={{
            ...getOneRandomModelMaxLength(),
            importProcessState: {
              id: "some-id",
              status: ImportProcessStateAPISpecs.Enums.Status.COMPLETED,
              result: {
                parsingErrors: false,
                parsingWarnings: false,
                errored: false,
              },
            },
          }}
        />
      );

      // WHEN the menu item is clicked
      fireEvent.click(screen.getByTestId(DATA_TEST_ID.MENU_ITEM_EXPORT));

      // THEN the notifyOnClose should be called to close the menu
      await waitFor(() => expect(stdGivenProps.notifyOnClose).toHaveBeenCalled());
      // AND the notifyOnExport should be called after the menu is closed
      expect(stdGivenProps.notifyOnExport).toHaveBeenCalledAfter(stdGivenProps.notifyOnClose as jest.Mock);
    });

    test("on close the notifyOnClose should be called", () => {
      // GIVEN the ContextMenu is rendered with the std given props
      render(<ContextMenu {...stdGivenProps} />);

      // WHEN the use closed the menu ( by pressing the escape key )
      const actualMenu = screen.getByTestId(DATA_TEST_ID.CONTEXT_MENU);
      fireEvent.keyDown(actualMenu, { key: "Escape" });

      // THEN the notifyOnClose should be called to close the menu
      expect(stdGivenProps.notifyOnClose).toHaveBeenCalled();
      // AND the notifyOnExport should not be called
      expect(stdGivenProps.notifyOnExport).not.toHaveBeenCalled();
    });
  });
});
