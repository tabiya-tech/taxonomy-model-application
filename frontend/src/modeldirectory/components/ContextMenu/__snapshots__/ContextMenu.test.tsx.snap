// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenu Render tests should render the ContextMenu visible 1`] = `
<div
  class="MuiPopover-root MuiMenu-root MuiModal-root css-10nakn3-MuiModal-root-MuiPopover-root-MuiMenu-root"
  data-testid="context-menu-715a535d-ca4b-4252-b967-8d5b5b2381b9"
  role="presentation"
>
  <div
    aria-hidden="true"
    class="MuiBackdrop-root MuiBackdrop-invisible MuiModal-backdrop css-g3hgs1-MuiBackdrop-root-MuiModal-backdrop"
    style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
  />
  <div
    data-testid="sentinelStart"
    tabindex="0"
  />
  <div
    class="MuiPaper-root MuiMenu-paper MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation2 MuiPopover-paper css-xem7h2-MuiPaper-root-MuiMenu-paper-MuiPopover-paper"
    style="opacity: 1; transform: scale(1, 1); transition: opacity 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 0ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; top: 16px; left: 16px; transform-origin: -16px -16px;"
    tabindex="-1"
  >
    <ul
      class="MuiList-root MuiList-padding MuiMenu-list css-6hp17o-MuiList-root-MuiMenu-list"
      role="menu"
      tabindex="-1"
    >
      <button
        class="MuiButtonBase-root MuiMenuItem-root MuiMenuItem-gutters Mui-focusVisible MuiMenuItem-root MuiMenuItem-gutters css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
        data-testid="menu-item-export-715a535d-ca4b-4252-b967-8d5b5b2381b9"
        role="menuitem"
        tabindex="0"
        type="button"
      >
        <div
          class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
            data-testid="CloudDownloadIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19.35 10.04C18.67 6.59 15.64 4 12 4 9.11 4 6.6 5.64 5.35 8.04 2.34 8.36 0 10.91 0 14c0 3.31 2.69 6 6 6h13c2.76 0 5-2.24 5-5 0-2.64-2.05-4.78-4.65-4.96zM17 13l-5 5-5-5h3V9h4v4h3z"
            />
          </svg>
        </div>
        <div
          class="MuiListItemText-root css-tlelie-MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
          >
            Export
          </span>
        </div>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </ul>
  </div>
  <div
    data-testid="sentinelEnd"
    tabindex="0"
  />
</div>
`;
